Userland/Libraries/LibWeb/HTML/NavigatorBeacon.cpp

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2023, Bastiaan van der Plaat <bastiaan.v.d.plaat@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibWeb/DOM/Document.h>
+#include <LibWeb/Fetch/Fetching/Fetching.h>
+#include <LibWeb/Fetch/Infrastructure/FetchAlgorithms.h>
+#include <LibWeb/HTML/Navigator.h>
+#include <LibWeb/HTML/NavigatorBeacon.h>
+#include <LibWeb/HTML/Window.h>
+
+namespace Web::HTML {
+
+// https://w3c.github.io/beacon/#sendbeacon-method
+WebIDL::ExceptionOr<bool> NavigatorBeaconMixin::send_beacon(String const& url, Optional<Fetch::BodyInit> const& data)
+{
+    auto& navigator = reinterpret_cast<Navigator&>(*this); // FIXME: This is ugly
+    auto& realm = navigator.realm();
+    auto& vm = realm.vm();
+    auto& relevant_settings_object = navigator.global_object().document()->relevant_settings_object();
+
+    // 1. Set base to this's relevant settings object's API base URL.
+    auto base_url = relevant_settings_object.api_base_url();
+
+    // 2. Set origin to this's relevant settings object's origin.
+    auto origin = relevant_settings_object.origin();
+
+    // 3. Set parsedUrl to the result of the URL parser steps with url and base. If the algorithm returns an error, or if parsedUrl's scheme is not "http" or "https", throw a "TypeError" exception and terminate these steps.
+    auto parsed_url = URLParser::basic_parse(url, base_url);
+    if (!parsed_url.is_valid())
+        return WebIDL::SimpleException { WebIDL::SimpleExceptionType::TypeError, MUST(String::formatted("Beacon URL {} is invalid.", url)) };
+    if (parsed_url.scheme() != "http" && parsed_url.scheme() != "https")
+        return WebIDL::SimpleException { WebIDL::SimpleExceptionType::TypeError, MUST(String::formatted("Beacon URL {} must be either http:// or https://.", url)) };
+
+    // 4. Let headerList be an empty list.
+    auto header_list = Fetch::Infrastructure::HeaderList::create(vm);
+
+    // 5. Let corsMode be "no-cors".
+    auto cors_mode = Fetch::Infrastructure::Request::Mode::NoCORS;
+
+    // 6. If data is not null:
+    Optional<JS::NonnullGCPtr<Fetch::Infrastructure::Body>> transmitted_data;
```

Comment:
`Optional<NonnullPtr>` ?  If we used a JS::GCPtr here, could we unconditionally set_body() below? or would that still need a null check.